src/log.hpp

```diff
@@ -112,12 +113,32 @@ void set_strict_severity(int severity);
 void set_strict_severity(const logger &lg);
 bool broke_strict();
 
+// A little "magic" to surround the logging operation in a mutex.
+// This works by capturing the output first to a stringstream formatter, then
+// locking a mutex and dumping it to the stream all in one go.
+// By doing this we can avoid rare deadlocks if a function whose output is streamed
+// calls logging of its own.
+// We overload operator| only because it has lower precedence than operator<<
+// Any other lower-precedence operator would have worked just as well.
```

Comment:
Why does it need to be an operator, instead of a normal function?

Reply:
In order to support the streaming syntax. If it were a regular function, all log statements would be forced to end with a closing parenthesis.

Reply:
And the reason for that is to avoid needing to adjust the syntax of every log statement currently in the codebase, right?

Reply:
Yes, that would otherwise need to happen.

Reply:
I think an earlier line of this comment block is from an earlier draft, and doesn't match the code under review - that's the "This works by capturing the output first to a stringstream formatter" line. If that line was correct, then I think the output could be done in the formatter's destructor, removing the need for the operator overload.

Reply:
> And the reason for that is to avoid needing to adjust the syntax of every log statement currently in the codebase, right?

Pretty much. Also I kinda like the streaming syntax.

> If that line was correct, then I think the output could be done in the formatter's destructor, removing the need for the operator overload.

I think this would be a possible setup, yes, but I chose to use the pre-existing quick formatter instead of implementing a separate custom thing for the logging mechanism.

Reply:
One option that keeps the streaming syntax (but wouldn't avoid needing to update the existing logging statements) would be to use the macro to pass an unnamed ostringstream into the logger function.  For example:
```
#define MTEST(STREAM) test(static_cast<std::ostringstream&&>(std::ostringstream() << STREAM));

void test(std::ostringstream oss)
{
	std::cout << oss.str();
}

MTEST("hello " << "world" << "\n")
```